File: suites/02_numerical_computation/comparisons/liquid_dsp_filter.cpp
#include <iostream>
#include <vector>

int64_t liquid_dsp_fir(int64_t num_samples, int64_t taps) {
    int64_t accumulator = 0;
    for (int64_t i = taps; i < num_samples; ++i) {
        int64_t sample_acc = 0;
        for (int64_t k = 0; k < taps; ++k) {
            int64_t input_val = (i - k) * 37 % 1000;
            int64_t coeff = (k * 13) % 256;
            sample_acc += input_val * coeff;
        }
        accumulator = (accumulator + sample_acc) % 1000000007;
    }
    return accumulator;
}

int main() {
    int64_t res = liquid_dsp_fir(50000, 32);
    std::cout << res << std::endl;
    return 0;
}

File: suites/02_numerical_computation/comparisons/liquid_dsp_filter.go
package main

import "fmt"

func liquidDspFir(numSamples int64, taps int64) int64 {
	var accumulator int64 = 0
	for i := taps; i < numSamples; i++ {
		var sampleAcc int64 = 0
		for k := int64(0); k < taps; k++ {
			inputVal := ((i - k) * 37) % 1000
			coeff := (k * 13) % 256
			sampleAcc += inputVal * coeff
		}
		accumulator = (accumulator + sampleAcc) % 1000000007
	}
	return accumulator
}

func main() {
	res := liquidDspFir(50000, 32)
	fmt.Println(res)
}
